shapes/main.cpp

```diff
@@ -1,54 +1,68 @@
+#include <algorithm>
+#include <functional>
 #include <iostream>
+#include <map>
+#include <memory>
 #include <vector>
-#include <algorithm>
 #include <string>
-#include <memory>
-#include "Shape.hpp"
+#include <type_traits>
+
+#include "Circle.hpp"
 #include "Rectangle.hpp"
+#include "Shape.hpp"
 #include "Square.hpp"
-#include "Circle.hpp"
 
 using namespace std;
 
 using Collection = vector<shared_ptr<Shape>>;
+using mapCollection = map<shared_ptr<Shape>, double>;
+using shapePtr = shared_ptr<Shape>;
 
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
-{
-    if(first == nullptr || second == nullptr)
-        return false;
+
+auto sortByArea = [](shapePtr first, shapePtr second) {
+    if(first == nullptr || second == nullptr) { return false; }
     return (first->getArea() < second->getArea());
-}
+};
 
-bool perimeterBiggerThan20(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getPerimeter() > 20);
+auto perimeterBiggerThan20 = [](shapePtr s) {
+    if (s) { return (s->getPerimeter() > 20); }
     return false;
-}
+};
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
+std::function<bool(shapePtr)> areaLessThanX = [x = 10](shapePtr s) {
```

Comment:
It can be auto here. The predicate parameter of the function findFirstShapeMatchingPredicate must be of type std::function.
```suggestion
auto areaLessThanX = [x = 10](shapePtr s) {
```

Reply:
done